source/multigrid/mg_tools.cc

```diff
@@ -1028,15 +1029,25 @@ namespace MGTools
 
     const unsigned int dofs_per_cell = dof.get_fe().n_dofs_per_cell();
     std::vector<types::global_dof_index> dofs_on_this_cell(dofs_per_cell);
+    std::vector<types::global_dof_index> cols;
+    cols.reserve(dofs_per_cell);
+
     for (const auto &cell : dof.cell_iterators_on_level(level))
       if (cell->is_locally_owned_on_level())
         {
           cell->get_mg_dof_indices(dofs_on_this_cell);
+          std::sort(dofs_on_this_cell.begin(), dofs_on_this_cell.end());
           for (unsigned int i = 0; i < dofs_per_cell; ++i)
-            for (unsigned int j = 0; j < dofs_per_cell; ++j)
-              if (mg_constrained_dofs.is_interface_matrix_entry(
-                    level, dofs_on_this_cell[i], dofs_on_this_cell[j]))
-                sparsity.add(dofs_on_this_cell[i], dofs_on_this_cell[j]);
+            {
+              cols.clear();
+              for (unsigned int j = 0; j < dofs_per_cell; ++j)
+                if (mg_constrained_dofs.is_interface_matrix_entry(
+                      level, dofs_on_this_cell[i], dofs_on_this_cell[j]))
+                  cols.push_back(dofs_on_this_cell[j]);
+              sparsity.add_row_entries(dofs_on_this_cell[i],
+                                       make_array_view(cols),
+                                       true);
```

Comment:
Would you mind making this consistent with the other functions and clear after adding entries,
```suggestion
              for (unsigned int j = 0; j < dofs_per_cell; ++j)
                if (mg_constrained_dofs.is_interface_matrix_entry(
                      level, dofs_on_this_cell[i], dofs_on_this_cell[j]))
                  cols.push_back(dofs_on_this_cell[j]);
              sparsity.add_row_entries(dofs_on_this_cell[i],
                                       make_array_view(cols),
                                       true);
              cols.clear();
```